README
TEACUP (TCP Experiment Automation Controlled Using Python) README 
=================================================================

$Id: README 1236 2015-04-14 07:51:18Z szander $

ACKNOWLEDGMENTS
---------------

TEACUP v0.9 was developed as part of a project funded by Cisco 
Systems and titled "Study in TCP Congestion Control Performance In 
A Data Centre". This is a collaborative effort between CAIA and 
Fred Baker of Cisco Systems.


INTRODUCTION
------------

This README is for TEACUP Version 0.9.

This README provides a brief overview on how to setup experiments.
It ONLY discusses the most commonly used config options.

For more details on how to setup experiments and the design of TEACUP
please refer to the technical report: 
http://caia.swin.edu.au/reports/150414A/CAIA-TR-150414A.pdf 

For more details on how to analyse the data collected during experiments
please refer to the technical report:
http://caia.swin.edu.au/reports/150414B/CAIA-TR-150414B.pdf

There also is a technical report that provides a command reference for
all tasks TEACUP provides:
http://caia.swin.edu.au/reports/150414C/CAIA-TR-150414C.pdf


TEACUP TESTBED
--------------

TEACUP makes some assumptions on the setup of the testbed and hosts. In order
to use all of TEACUP's functions one needs to install a testbed as described
briefly in INSTALL and described in more detail in:
http://caia.swin.edu.au/reports/150210C/CAIA-TR-150210C.pdf 

A minimum setup for TEACUP does not require all the functionality
described in CAIA-TR-150210C.pdf. The minimum setup is:
- Two subnets separated by one router, with at least one host in each
  subnet acting as traffic source/sink;
- Hosts must run Linux, FreeBSD, Windows 7 or Mac OS X, the router must run
  Linux (preferably) or FreeBSD;
- Each host needs an extra network interface connected to a separate control 
  network;
- A control host that runs TEACUP to control the testbed via the control
  network;
- TCP logging tools for OS must be installed as explained in CAIA-TR-150210C.pdf;
- Traffic generators must be installed as explained in CAIA-TR-150210C.pdf;
- Support tools must be installed as explained in CAIA-TR-150210C.pdf, but
  depending on the OS many of these are already installed by default.

A multi-boot OS setup with PXE booting and power controllers as decribed in
CAIA-TR-150210C.pdf is very useful but optional.


CONFIG FILE
-----------

Copy config.py.example to config.pt in your experiment directory and modify
it. The following discusses the most important options.

1. Fabric configuration

The following settings in the config file are Fabric settings. For a 
more in-depth description refer to the Fabric documentation. All Fabric 
settings are part of the Fabric env dictionary and hence Python variables 
(and must adhere to the Python syntax).

env.user = 'root'

The user used for the SSH logins.

env.password = 'rootpw'

The password used for the SSH logins. This can be empty if public-key 
authorisation is set up properly,

2. Testbed configuration

All settings start with the TPCONF_ prefix and are Python variables (and 
must adhere to the Python syntax).

TPCONF_script_path = '/home/<user>/src/teacup-0.9'
sys.path.append(TPCONF_script_path)

The path were the TEACUP scripts are located. This is appended to the Python 
path (DO NOT remove the second line).

TPCONF_router = [ 'testrouter', ]
TPCONF_hosts = [ 'testhost1', 'testhost2', ]

Two lists specify the testbed hosts. TPCONF_router specifies the list of 
routers. Note that currently the TPCONF_router list is limited to only _one_ 
router. TPCONF_hosts specifies the list of hosts. The hosts can be specified 
as IP addresses or host names (just the names of fully qualified domain names).

TPCONF_host_internal_ip = {
        'testrouter' :  [ '172.16.10.1', '172.16.11.1' ],
        'testhost1' : [ '172.16.10.2' ],
        'testhost2' : [ '172.16.10.3' ],
}

This dictionary specifies the internal addresses / interfaces for each host. The 
hosts (keys) specified must match the entries in the TPCONF_router and TPCONF_hosts 
lists exactly. The current code does simple string matching, it
does NOT attempt to resolve host identifiers in some canonical form.

3. Experiment general settings

now = datetime.datetime.today()
TPCONF_test_id = now.strftime("%Y%m%d-%H%M%S") + '_experiment'

TPCONF_test_id specifies the default test ID prefix. Note that if the test ID is 
specified on the command line, the command line overrules this setting.

TPCONF_host_os = {
        'testrouter' : 'Linux',
        'testhost1'  : 'FreeBSD',
        'testhost2'  : 'Linux',
}

The TPCONF_host_os dictionary specifies which OS are booted on the different hosts. 
The hosts (keys) specified must match the entries in the TPCONF_router and TPCONF_hosts 
lists exactly. The current code does simple string matching, it does _not_ attempt to 
resolve host identifiers in some canonical form.

The three different types of OS supported are: Linux, FreeBSD and CYGWIN (Windows). 
Selecting the specific Linux kernel to boot is not supported yet (the name of the 
kernel is hard-coded).

TPCONF_force_reboot = '1'

If TPCONF_force_reboot is set to '1' _all_ hosts will be rebooted. If 
TPCONF_force_reboot is set to '0' only hosts where the currently running OS is NOT the 
desired OS (specified in TPCONF_host_os) will be rebooted.

For the automatic OS selection to work, the testbed must have a PXE-based setup as
described in http://caia.swin.edu.au/reports/150210C/CAIA-TR-150210C.pdf

4. Experiment router queues

TPCONF_router_queues = [
        # This example sets same delay/loss for every host, and same delay/loss 
        # in both directions
        ( '1', " source='172.16.10.0/24', dest='172.16.11.0/24', delay=V_delay, "
               "loss=V_loss, rate=V_up_rate, queue_disc=V_aqm, queue_size=V_bsize " ),
        ( '2', " source='172.16.11.0/24', dest='172.16.10.0/24', delay=V_delay, "
               "loss=V_loss, rate=V_down_rate, queue_disc=V_aqm, queue_size=V_bsize " ),
]

The TPCONF_router_queues specifies the router queues. Each entry is a 2-tuple. The 
first value specifies a unique integer ID for each queue. The second value is a 
comma-separated string specifying the queue parameters.

The queues do not necessarily need to be defined in the order of queue ID but it is 
recommended to do so.

The following queue parameters exist:

source: 
Specifies the source IP / hostname or source network (<ip|hostname>[/<prefix>]) of 
traffic that is queued in this queue. If a host name is specified there can be no prefix. 
One can specify an internal/testbed or external/control IP/hostname. If an external 
IP/hostname is specified this will be automatically translated into the first internal 
IP specified for the host in TPCONF_host_internal_ip.

dest:
Specifies the destination IP / hostname or source network (<ip|hostname>[/<prefix>]) of 
traffic that is queued in this queue. If a host name is specified there can be no prefix. 
One can specify an internal/testbed or external/control IP/hostname. If an external 
IP/hostname is specified this will be automatically translated into the first internal IP 
specified for the host in TPCONF_host_internal_ip.

delay:
Specifies the emulated constant delay in milliseconds. For example, delay=50 sets the 
delay to 50ms.

loss:
Specifies the emulated constant loss rate. For example, loss=0.01 sets the loss rate to 1%.

rate:
Specifies the rate limit of the queue. On Linux we can use units such as kbit or mbit. 
For example, queue_size=1mbit sets the rate limit to 1 Mbit/second.

queue_size:
Specifies the size of the queue. On Linux queue size is usually defined in packets, 
only for some queuing disciplines we can specify the size in bytes. For example, 
queue_size=1000 sets the queue size to 1000 packets. On FreeBSD the queue size is also
specified in packets typically, but one specify the size in bytes by adding a 'bytes' or 
kbytes', for example queue_size=100kbytes specifies a queue of size 100kbytes. If 'bdp' is 
specified the queue size will be set to the nominal bandwidth-delay-product (this does only 
work for queuing disciplines where we know whether the queue limit is in bytes or packets).

The minimum queue size is 1 packet (if the limit is specified in packets) or 2048bytes 
(if the limit is specified in bytes).

queue_disc:
Specifies the queuing discipline. This can be the same of any of the queuing disciplines 
supported by Linux, such as 'fq_codel', 'codel', 'red', 'choke', 'pfifo', 'pie' etc. On 
FreeBSD the only queuing discipline available are 'fifo' and 'red'. For example, 
queue_disc='fq_codel' sets the queuing discipline to the fair-queuing+codel model. For 
compatibility, with FreeBSD on can specify 'fifo' on Linux, which is mapped to 'pfifo' 
('pfifo' is the default for HTB classes, we use HTB for rate limiting). The queue_disc 
parameter must be specified explicitly.

queue_disc_params:
This allows to pass parameters to queuing disciplines. For example, if we wanted to turn 
ECN on for fq_codel we would specify queue_disc_params='ecn' (see fq_codel man page).

bidir:
This allows to specify whether a queue is unidirectional (set to '0') or bidirectional 
(set to '1'). A unidirectional queue will only get the traffic from source to dest, 
whereas a bidirectional queue will get the traffic from source to dest AND from dest 
to source.

All parameters must be assigned with either a constant value or an variable name 
(V_<name>). Variable names must be defined in TPCONF_parameter_list and/or 
TPCONF_variable_defaults. Variables are replaced with either the default
value specified in TPCONF_variable_defaults or the current value from TPCONF_parameter_list 
if we iterate through multiple values for the parameter.

5. Experiment traffic generators

Traffic generators are defined by the variable TPCONF_traffic_gens. This is a list of 
3-tuples. The first value of a tuple is the start time relative to the start of the experiment. 
The second value of the tuple is the a unique ID. The third value of the tuple is a string 
containing the function name of the start function of the generator as first name followed 
by the parameters.

Client and server parameters can be external addresses or host names. An external address 
or host name is replaced by the first internal address specified for a host in 
TPCONF_host_internal_ip. Client and server parameters can also be internal addresses, which
allows to specify any internal address existing.

Each parameter is defined as <parameter_name>=<parameter_value>. Parameter names must be 
the parameter names of traffic generator functions (and as such be valid Python variable 
names). Parameter values can be either constants (string or numeric) or
variables that are replaced by the actual values depending on the current experiment. These 
variables must be named V_<name> and they must be defined in either TPCONF_parameter_list 
or TPCONF_variable_defaults.

The following shows a simple example. At time zero a web server is started and fake DASH 
content is created. 0.5 seconds later a httperf client in DASH mode is started. The duration 
and rate of the DASH-like flow are specified by variables, which can change for each experiment. 
In contrast the cycle length and prefetch time are set to fixed values.

TPCONF_traffic_gens = [
        ( '0.0', '1', " start_http_server, server='testhost1', port=80 " ),
        ( '0.0', '2', " create_http_dash_content, server='testhost1', duration=2*V_duration, "
                      "rates=V_dash_rates, cycles='5, 10' " ),
        ( '0.5', '3', " start_httperf_dash, client='testhost2', server='testhost1', port=80, "
                      "duration=V_duration, rate=V_dash_rate, cycle=5, prefetch=15 " ),

TEACUP supports the following traffic generators:
- TCP bulk transfer (iperf)
- HTTP traffic (httperf)
- TCP video streaming (httperf)
- Incast traffic (one querier, N responders) (httperf)
- ping
- UDP (iperf)
- VoIP-like (fixed packet sizes and interval) (nttcp)

Available traffic generators and their parameters are described in:
http://caia.swin.edu.au/reports/150414A/CAIA-TR-150414A.pdf

6. Experiment variables

There are two types of variables: singulars and lists. Singulars are fixed parameters
while lists specify the different values used in subsequent experiments based on the 
definitions of TPCONF_parameter_list and TPCONF_vary_parameters (see below).

TPCONF_duration = 30

The duration of the experiment in seconds (must be an integer).

TPCONF_runs = 1

The number of runs carried out for each unique parameter combination.

TPCONF_TCP_algos = [ 'newreno', 'cubic', 'htcp', ]

The TCP congestion algorithms used. The following algorithms can be selected: 'newreno', 
'cubic',  'htcp', 'cdg', 'compound'. However, only some of these are supported depending 
on the OS a host is running:
Windows: newreno (default), compound
FreeBSD: newreno (default), cubic, htcp, cdg
Linux: cubic (default), newreno, htcp

Instead of specifying a particular TCP algorithm one can specify 'default'. This will 
set the  algorithm to the default algorithm depending on the OS the host is running (see 
above).

Using only TPCONF_TCP_algos one is limited to either using the same algorithm on all hosts or 
the defaults. To run different algorithms on different hosts, one can specify 'host<N>' where 
<N> is an integer number starting from 0. The <N> refers to the N-th entry for each host in 
TPCONF_host_TCP_algos.

TPCONF_host_TCP_algos = {
        'testhost1' : [ 'default', 'newreno', ],
        'testhost2' : [ 'default', 'newreno', ],
}

This defines the TCP congestion control algorithms used for each host if the 'host<N>' 
definitions are used in TPCONF_TCP_algos. In the above example a 'host0' entry in 
TPCONF_TCP_algos will lead to each host using its default. A 'host1' entry will set
both hosts to use 'newreno'. 

TPCONF_delays = [ 0, 25, 50, 100 ]

The emulated delays in milliseconds.

TPCONF_loss_rates = [ 0, 0.001, 0.01 ]

The emulated loss rates. The numbers must be between 0.0 and 1.0.

TPCONF_bandwidths = [
        (  '8mbit', '1mbit'   ),
        ( '20mbit', '1.4mbit' ),
]

The emulated bandwidths as 2-tuples. The first value in each tuple is the downstream 
rate and the second value in each tuple is the upstream rate. Note that the values are 
passed through to the router queues and are not parsed. Units can be used if the queue setup 
allows this, e.g. above we use mbit to specify Mbits/second which Linux tc framework allows 
us to do.

TPCONF_aqms = [ 'pfifo', 'fq_codel', 'pie' ]

The list of AQM/queuing techniques. This is completely dependent on the router OS. Linux 
supports  'pfifo', 'bfifo', 'fq_codel', 'codel', 'pie', 'red', ... (see tc man page for full 
list). FreeBSD support only 'fifo' and 'red'. Default on Linux and FreeBSD are FIFOs.

TPCONF_buffer_sizes = [ 100, 200 ]

If router is Linux this is mostly in packets/slots, but it depends on AQM (e.g. for bfifo 
it's  bytes). If the router is FreeBSD this would be in slots by default, but we can specify 
byte sizes (e.g. we can specify 4Kbytes).

TPCONF_vary_parameters = [ 'tcpalgos', 'delays', 'loss', 'bandwidths', 'aqms', 'bsizes', 
     'runs', ]

This specifies this list of parameters we vary, i.e. parameters that have multiple values. 
These parameter must be defined in TPCONF_parameter_list. The total number of experiments 
carried out is the number of unique parameter combinations multiplied by the number of 
TPCONF_runs if 'runs' is specified here. This is also the order of the parameters in the 
file names. While not strictly necessary 'runs' should be last in the list (if used). If 
'runs' is not specified there is a single experiment for each parameter combination.

TPCONF_vary_parameters is only used for multiple experiments, i.e. 
'fab run_experiment_multiple' (see below). When we run 'fab run_experiment_single' all 
the variables are set to fixed values based on TPCONF_variable_defaults. 
TPCONF_parameter_list specifies the variable parameters.

TPCONF_variable_defaults and TPCONF_parameter_list are explained in detail in:
http://caia.swin.edu.au/reports/150414A/CAIA-TR-150414A.pdf

7. Example with two variables

Let's discuss a simple example where we vary delay and TCP algorithms. Assume we want to 
experiment with two delay settings and two different TCP CC algorithms. So we have:

TPCONF_delays = [ 0, 50 ]
TPCONF_TCP_algos = [ 'newreno', 'cubic' ]

We also need to specify the two parameters to be varied :

TPCONF_parameter_list = {
        'delays'        :  ( [ 'V_delay' ],        [ 'del' ],    TPCONF_delays,        {} ),
        'tcpalgos'      :  ( [ 'V_tcp_cc_algo' ],  [ 'tcp' ],    TPCONF_TCP_algos,     {} ),
 	OTHERS OMITTED	

TPCONF_vary_parameters = [ 'delays', 'tcpalgos' ]

V_delay can then be used in the router queue settings. V_tcp_cc_algo is passed to the host setup 
function. When we run 'fab run_experiment_multiple' this will run the following experiments, 
here represented by the start of the log file names. (We assume the test ID prefix is the 
default and the experiment was run on the 6th of Dec 2014.)


ChangeLog
Version 0.9.3 (6th May 2015)
----------------------------
- Fixed wrong pdf file names for some analysis functions when multiple test 
  IDs are specified
- Extended init_os, so we can reboot non-resposnsive host(s) into a desired
  operating system. User must specify MAC addresses of control/boot interfaces
  of the hosts with the new parameter mac_list. Of course this only works if
  a supported power controller is installed and configured for teacup, so 
  teacup can power cycle the host(s).
- Fixed bug in analyse functions that caused some flows not to be plotted
  when function was called with a list of more than one test id
- Added packet loss analysis function (ONLY works for game traffic thus far)
- Fixed bug that plotted the same flow twice in one graph, in case we specify 
  two (or more) source filters with the same IP and port (irrespective of 
  whether they are S_ or D_ filters)
- Added FILTER_FLOWS environment variable for time series plots. if set to 1, 
  all flows that were not active will be filtered out. maybe this should be 
  the default behaviour in the future, for now I have added it via the variable.

Version 0.9.2 (28th April 2015)
-------------------------------
- Fixed extract_rtt and analyse_rtt so that if burst_sep = 0 then
  timestamps are not normalised before plotting for each flow individually,
  but plotting will normalise all timestamps based on the timestamp of the
  first packet during the experiments. Basically this restires the behaviour
  we had pre version 0.9. 
- Fixed extract_ackseq and analyse_ackseq accordingly (see last point)
- Added tags in code and script to generate code documentation with doxygen
- For analyse_2d_density replaced merge_xdata and merge_ydata with merge_data
- For burst-based analysis, instead of starting every burst at time zero,
  we now normalise the start time of burst n of flow f based on the start 
  time of the first burst n over all flows. this means if burst 1 of flow
  a start a bit later than burst 1 of flow b, we now see the gap in time. 
- Fixed bug in extract_rtt/analyse_rtt that causes wrong results if there
  are multiple flows between the same two hosts with the same server port
- Added client start delay parameter for game traffic generator
- Changed type of plot from "both" to "points" for analyse_dash_goodput
  so it is consistent with the other plot functions

Version 0.9.1 (17th April 2015)
-------------------------------
- Fixed bug in caching of hosts that participated in experiment
  (only occurs if processing experiments done with pre 0.9 versions)
- Fixed bug in extract_rtt that caused incorrect RTT estimates in plots

Version 0.9 (14th April 2015)
-----------------------------
- Fixed broken ack seq data extract with more than one test id
- Fixed broken/missing burst_seq type conversion in analyse_ackseq
- Added link speed setting for testbed interfaces on all hosts (except router).
  Speed can be set to 10, 100, 100 or auto (default) 
- added extract and plot method for incast inter-query times 
  (analyse_incast_iqtimes)
- added extract method for incast response times from tcpdump data
- Fixed broken ts_correct for analyse_rtt
- Timestamp correction is now enabled by default for all analysis functions
  (ts_correct=1). If no clock offsets file can be computed, e.g. because
  multicast ping was not enabled, a warning is printed and uncorrected 
  timestamps are used.
- Added logging for link speed setting on Windows
- Added response time as metric for analyse_cmpexp
- Added merge_data parameter for analsye_cmpexp that if set to '1'
  will merge data from different flows/responders into one dataset
- Added merge_xdata and merge_ydata parameters for analyse_2_density. if set
  to '1' will merge data from different flows/responders into one dataset for
  x-axis or y-axis data
- Removed defaults for variables for init_topology. Must set command line
  parameters or TPCONF variables. This is to protect against accidentally
  configuring wrong switch.
- Added multiple router support. TPCONF_router_queues can now be a dictionary
  indexed by router names (listed in TPCONF_router) and the values are
  of the same structure as TPCONF_router_queues. For backwards compatibility
  TPCONF_router_queues can also be defined as list of tuples as before.
- Optimised file finding with an experiment ID to directory cache (and also
  removed a superfluous "find" in clock offset correct function). The cache
  file is called teacup_dir_cache.txt and is created in the directory where
  fab is executed. Cache entries are never deleted, so if one renames directories
  or moves files, one should clear the whole cache by deleting the cache file 
  or by selectively removing entries from the cache file.
- allow absolute paths for out_dir, res_dir and pdf_dir
- removed local('which <executable>') tests in analyse.py
- added a way to specify the config file used. It can be specified using
  fab's --set parameter to set the env variable teacup_config as follows:
  fab --set teacup_config=config.py 
- fixed crash in analyse_incast_iqtimes when ts_correct=1
- added diff_to_burst_start parameter to analyse_incast_iqtimes, which allows 
  to select whether time gaps are from the start of bursts or between consecutive 
  packets
- support for iqtimes in analyse_cmpexp, limited to time gaps between
  consecutive packets. also data needs to be extracted beforhand as there
  is no query_host parameter for analyse_cmpexp yet.
- added sburst and eburst parameters for extract_incast and analyse_incast
  (not for tcpdump data though, only for httperf log extraction)
- added sburst and eburst parameters for analyse_cmpexp, so can be used
  for response times or ackseq
- fixed bug in analyse_cmpexp when group_by_prefix=1 that caused duplicate
  x-axis labels and the plot function to fail
- all varying parameters for a series of experiments are now logged in
  <test_id_pfx>_varying_params.log.gz files

fabfile.py
# Copyright (c) 2013-2015 Centre for Advanced Internet Architectures,
# Swinburne University of Technology. All rights reserved.
#
# Author: Sebastian Zander (szander@swin.edu.au)
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
## @package fabfile
# fabfile
#
# $Id: fabfile.py 1302 2015-05-01 07:47:33Z szander $

import os
import imp
import datetime
from fabric.api import execute, task, env, abort, local

# this will print paramiko errors on stderr
import logging
logging.getLogger('paramiko.transport').addHandler(logging.StreamHandler())

# if the env variable teacup_config is set import the config file specified.
# the variable is set as follows: fab --set teacup_config=myconfig.py
# otherwise import the default config.py
try:
    if not os.path.isfile(env.teacup_config):
        abort('Specified config file %s does not exit' % env.teacup_config) 

    # before loading config change to the directory of specified config file
    # this is needed to make execfile without absolute path in config file work
    # afterwards change back directory to current path
    curr_dir = os.getcwd()
    config_dir = os.path.dirname(env.teacup_config)
    if config_dir == '':
        config_dir = '.'
    os.chdir(config_dir)
    config = imp.load_source('config', env.teacup_config)
    os.chdir(curr_dir)

except AttributeError:
    import config

# now import all the teacup modules. to be backwards compatible with old
# versions (since we can select the configuration file now), we MUST wrap
# all new tasks in try-except. we can group together tasks that were added
# as batch, i.e. if they were all added together in one version, they
# can be in the same try clause

from experiment import run_experiment
from hosttype import get_type
from hostint import get_netint
from hostmac import get_netmac
from sanitychecks import check_host, check_connectivity, kill_old_processes, \
    sanity_checks, get_host_info, check_config, check_time_sync
from util import exec_cmd, authorize_key, copy_file
from hostsetup import init_host, init_ecn, init_cc_algo, init_router, \
        init_hosts, init_os, power_cycle, init_host_custom

try:
    from hostsetup import init_topology
except ImportError:
    pass

from loggers import start_tcpdump, stop_tcpdump, start_tcp_logger, \
    stop_tcp_logger, start_loggers, log_sysdata, log_queue_stats, \
    log_config_params, log_host_tcp, log_varying_params
from routersetup import init_pipe, show_pipes
from trafficgens import start_iperf, start_ping, \
    start_http_server, start_httperf, \
    start_httperf_dash, create_http_dash_content, \
    create_http_incast_content, start_httperf_incast

try:
    from trafficgens import start_httperf_incast_n 
except ImportError:
    pass

from analyse import analyse_rtt, analyse_cwnd, analyse_tcp_rtt, \
        analyse_throughput, analyse_all, analyse_dash_goodput, \
        analyse_tcp_stat, analyse_cmpexp

try:
    from analyse import analyse_incast
except ImportError:
    pass

try:
    from analyse import extract_rtt, extract_cwnd, extract_tcp_rtt, \
        extract_all, extract_dash_goodput, \
        extract_tcp_stat, extract_incast
except ImportError:
    pass

try:
    from analyse import extract_throughput
except ImportError:
    pass

# renamed extract_throughput to extract_pktsizes in version 0.9
try:
    from analyse import extract_pktsizes
except ImportError:
    pass

try:
    from analyse import analyse_ackseq, extract_ackseq
except ImportError:
    pass

try:
    from analyse import analyse_2d_density, extract_incast_iqtimes, \
         extract_incast_restimes, analyse_incast_iqtimes, analyse_goodput
except ImportError:
    pass

try:
    from analyse import extract_pktloss, analyse_pktloss
except ImportError:
    pass

try:
    from clockoffset import get_clock_offsets, adjust_timestamps
except ImportError:
    pass

try:
    pass # initialsetup is not public
except ImportError:
    pass

try:
    from version import get_version 
except ImportError:
    pass

try:
    from backupconfig import backup_config, dump_config_vars
except ImportError:
    pass

try:
    from nameipmap import get_nameip_map
except ImportError:
    pass

try:
    from internalutil import mkdir_p
except ImportError:
    pass


## Set to zero if we don't need OS initialisation anymore
# XXX this is a bit ugly as a global
do_init_os = '1'

## List of experiment completed (from experiments_completed.txt)
experiments_done = {}


## Set all basic parameters not yet set to single values
#  @param nargs Arguments
#  @param kwargs Keyword arguments
#  @return Extended nargs, kwargs
def _fill_missing(*nargs, **kwargs):

    global do_init_os

    for k, v in config.TPCONF_variable_defaults.items():
        if k not in kwargs:
            kwargs[k] = v

    # for compatibility with internal parameters
    if 'V_ecn' in kwargs:
    	kwargs['ecn'] = kwargs['V_ecn']
    if 'V_tcp_cc_algo' in kwargs:
        kwargs['tcp_cc_algo'] = kwargs['V_tcp_cc_algo']
    kwargs['duration'] = kwargs['V_duration']

    # set special defaults
    if 'run' not in kwargs:
        kwargs['run'] = 0
    if 'do_init_os' not in kwargs:
        kwargs['do_init_os'] = do_init_os

    return nargs, kwargs


## Check if experiment has been done before based on
## experiments_completed.txt file
#  @param test_id Experiment ID
#  @return True if experiment has been done before, false otherwise
def _experiment_done(test_id=''):
    global experiments_done

    if len(experiments_done) == 0:
        try:
            with open('experiments_completed.txt') as f:
                experiments = f.readlines()

            for experiment in experiments:
                experiments_done[experiment.strip()] = 1
        except IOError:
            # don't care if experiments_completed.txt does not exist, it
            # will be created after the next experiment
            pass

    if experiments_done.get(test_id, 0) == 0:
        return False
    else:
        return True


## Check config file and log config information
#  @param test_id Test ID
def config_check_and_log(test_id):
    execute(check_config, hosts=['MAIN'])  # use a dummy host here and below

    # create sub directory for test id prefix
    mkdir_p(test_id)
    execute(log_varying_params, test_id, test_id, hosts=['MAIN'])

    # make copy of config and dump TPCONF variables
    # wrap these in try-except cause they don't exist for old versions
    try:
        execute(backup_config, test_id, hosts=['MAIN'])
        execute(dump_config_vars, test_id, hosts=['MAIN'])
        execute(get_nameip_map, test_id, hosts=['MAIN'])
    except:
        pass


## Run single experiment (TASK)
#  @param test_id Test ID prefix
#  @param nargs Arguments (user-supplied parameters)
#  @param kwargs Keyword arguments (user-supplied parameters)
@task
def run_experiment_single(test_id='', *nargs, **kwargs):
    "Run a single experiment"

    if test_id == '':
        test_id = config.TPCONF_test_id

    config_check_and_log(test_id)
    
    _nargs, _kwargs = _fill_missing(*nargs, **kwargs)
    execute(run_experiment, test_id, test_id, *_nargs, **_kwargs)


## Generic function for varying a parameter
#  @param test_id Test ID
#  @param test_id_pfx Test ID prefix
#  @param resume '0' do all experiment, '1' do not repeat experiment if done according
#                to experiments_completed.txt
#  @param var_list List of parameters to vary
#  @param names List of variable names corrsponding to parameters
#  @param short_names List of short names used in file names corrsponding to parameters
#  @param val_list List of variable values corrsponding to parameters
#  @param extra_params Extra variables we need to set
#  @param nargs: variables we set and finally pass to run_experiment
#  @param kwargs: variables we set and finally pass to run_experiment
def _generic_var(test_id='', test_id_pfx='', resume='0', var_list=[], names=[
], short_names=[], val_list=[], extra_params={}, *nargs, **kwargs):

    global do_init_os

    # set if not set yet
    if test_id == '':
        test_id = config.TPCONF_test_id
    if test_id_pfx == '':
        test_id_pfx = test_id

    # remove current variable
    var_list.pop(0)

    for k, v in extra_params.items():
        kwargs[k] = v

    for val in val_list:
        if len(names) == 1:
            # add parameter and parameter value to test_id
            _test_id = test_id + '_' + \
                short_names[0] + '_' + str(val).replace('_', '-')
            # push value on parameter list
            kwargs[names[0]] = val
        else:
            _test_id = test_id
            c = 0
            for name in names:
                # add parameter and parameter value to test_id
                _test_id = _test_id + '_' + \
                    short_names[c] + '_' + str(val[c]).replace('_', '-')
                # push value on parameter list
                kwargs[name] = val[c]
                c += 1

        if len(var_list) > 0:
            # if we have another parameter to vary call the appropriate
            # function
            next_var = var_list[0]
            _names, _short_names, _val_list, _extra_params = config.TPCONF_parameter_list[
                next_var]
            # important that we pass a copy of var_list here and not a
            # reference
            _generic_var(
                _test_id,
                test_id_pfx,
                resume,
                list(var_list),
                _names,
                _short_names,
                _val_list,
                _extra_params,
                *nargs,
                **kwargs)
        else:
            # else fill in any missing parameters and start experiment
            _nargs, _kwargs = _fill_missing(*nargs, **kwargs)
            if resume == '0' or not _experiment_done(_test_id):
                #print('run', _test_id, _nargs, _kwargs)
                execute(
                    run_experiment,
                    _test_id,
                    test_id_pfx,
                    *_nargs,
                    **_kwargs)
                do_init_os = '0'


## Run a series of experiments varying different things (TASK)
#  @param test_id Test ID prefix
#  @param resume '0' do all experiment, '1' do not repeat experiment if done
#                according to experiments_completed.txt
#  @param nargs Various parameters (supplied by user)
#  @param kwargs Various parameters (supplied by user)
@task
def run_experiment_multiple(test_id='', resume='0', *nargs, **kwargs):
    "Run series of experiments"

    if test_id == '':
        test_id = config.TPCONF_test_id

    config_check_and_log(test_id)

    var_list = config.TPCONF_vary_parameters

    if len(var_list) > 0:
        # if we have another parameter to vary call the appropriate function
        next_var = var_list[0]
        names, short_names, val_list, extra_params = config.TPCONF_parameter_list[
            next_var]
        # important that we pass a copy of var_list here and not a reference
        _generic_var(
            test_id,
            test_id,
            resume,
            list(var_list),
            names,
            short_names,
            val_list,
            extra_params,
            *nargs,
            **kwargs)
    else:
        # else fill in any missing parameters and start experiment
        _nargs, _kwargs = _fill_missing(*nargs, **kwargs)
        if resume == '0' or not _experiment_done(test_id):
            #print('run', test_id, _nargs, _kwargs)
            execute(run_experiment, test_id, test_id, *_nargs, **_kwargs)
            do_init_os = '0'
